MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+                       GeoLib::LineSegment const& seg1)
+            {
+                // min dist for segment 0
+                const double d0(
+                    std::min(MathLib::sqrDist(beg_pnt, seg0.getBeginPoint()),
+                             MathLib::sqrDist(beg_pnt, seg0.getEndPoint())));
+                // min dist for segment 1
+                const double d1(
+                    std::min(MathLib::sqrDist(beg_pnt, seg1.getBeginPoint()),
+                             MathLib::sqrDist(beg_pnt, seg1.getEndPoint())));
+                return d0 < d1;
+            });
+        GeoLib::Point * pnt{
+            MathLib::sqrDist(beg_pnt, min_dist_segment->getBeginPoint()) <
+                    MathLib::sqrDist(beg_pnt, min_dist_segment->getEndPoint())
+                ? new GeoLib::Point{min_dist_segment->getBeginPoint()}
+                : new GeoLib::Point{min_dist_segment->getEndPoint()}};
+        sub_segments.emplace_back(
+            GeoLib::LineSegment{new GeoLib::Point{beg_pnt, 0}, pnt, true});
+    }
+    // sort all sub segments for the given segment (beg_pnt, end_pnt)
+    GeoLib::sortSegments(beg_pnt, sub_segments);
+
+    sub_segments.erase(std::unique(sub_segments.begin(), sub_segments.end()),
+                       sub_segments.end());
+
+    return sub_segments;
+}
+
+static void mapPointOnSurfaceElement(MeshLib::Element const& elem,
+                                     MathLib::Point3d& q)
+{
+    // create plane equation: n*p = d
+    MathLib::Vector3 const& p(*(elem.getNode(0)));
+    MathLib::Vector3 const n(MeshLib::FaceRule::getSurfaceNormal(&elem));
+    if (n[2] == 0.0) { // vertical plane, z coordinate is arbitrary
+        q[2] = p[2];
+    } else {
+        double const d(MathLib::scalarProduct(n, p));
+        q[2] = (d - n[0]*q[0] - n[1]*q[1])/n[2];
+    }
+}
+
+static std::vector<MeshLib::Element const*>
+getCandidateElementsForLineSegmentIntersection(
+    MeshLib::MeshElementGrid const& mesh_element_grid,
+    GeoLib::LineSegment segment)
+{
+    // modify z coordinates such that all surface elements around the line
+    // segment are found
+    segment.getBeginPoint()[2] = mesh_element_grid.getMinPoint()[2];
+    segment.getEndPoint()[2] = mesh_element_grid.getMaxPoint()[2];
+    std::array<MathLib::Point3d, 2> const pnts{
+        {segment.getBeginPoint(), segment.getEndPoint()}};
+    GeoLib::AABB aabb(pnts.cbegin(), pnts.cend());
+
+    auto candidate_elements = mesh_element_grid.getElementsInVolume(
+        aabb.getMinPoint(), aabb.getMaxPoint());
+
+    // make candidate elements unique
+    BaseLib::makeVectorUnique(candidate_elements);
+
+    return candidate_elements;
+}
+
+static bool snapPointToElementNode(MathLib::Point3d& p,
+                                   MeshLib::Element const& elem, double eps)
+{
+    for (std::size_t k(0); k<elem.getNumberOfNodes(); ++k) {
+        double const sqr_dist_2d(MathLib::sqrDist2d(p, *(elem.getNode(k))));
+        if (sqr_dist_2d < eps) {
```

Comment:
the epsilon here (and in other places) is given for the squared distance. Is that consistent everywhere? Personally, I'd expect that tolerances are given w.r.t. the distance, not to its square.